impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -401,26 +491,34 @@ private void invalidate() {
       
       private void advance() {
         next = null;
-        while (next == null && it.hasNext()) {
-          Map.Entry<K, OnHeapValueHolder<V>> entry = it.next();
-          final long now = timeSource.getTimeMillis();
-          if (entry.getValue().isExpired(now, TimeUnit.MILLISECONDS)) {
-            it.remove();
-            onExpiration(entry.getKey(), entry.getValue());
-            continue;
+        try {
+          while (next == null && it.hasNext()) {
+            Map.Entry<K, OnHeapValueHolder<V>> entry = it.next();
+            final long now = timeSource.getTimeMillis();
+            if (entry.getValue().isExpired(now, TimeUnit.MILLISECONDS)) {
+              it.remove();
+              onExpiration(entry.getKey(), entry.getValue());
+              continue;
+            }
+
+            next = entry;
           }
-          
-          next = entry;
+        } catch (RuntimeException re) {
+          prefetchFailure = new CacheAccessException(re);
         }
       }
       
       @Override
-      public boolean hasNext() throws CacheAccessException {
+      public boolean hasNext() {
         return next != null;
```

Comment:
According to my reading of the code, if you have a `prefetchFailure`, then `next` is null. So in order to have `hasNext()` return true, it should check both `next` and `prefetchFailure` and either different than `null` indicates there is/was something ... a real one or an exception waiting to happen.